File: ui/contracts/src/framing.ts
/// Length-prefixed frame codec mirroring `runtime/ipc/include/mirage/runtime/ipc/framing.hpp`
/// and `framing.cpp`: 4-byte little-endian unsigned payload length followed by
/// the payload bytes, 1 MiB payload cap, no silent truncation. Consumed by the
/// browser/devbridge side once real transports land (M1.5-03/M1.5-05); kept in
/// the contract package so both ends share the exact framing rules.

export const MAX_FRAME_BYTES = 1024 * 1024;
export const FRAME_HEADER_BYTES = 4;

const textEncoder = new TextEncoder();
const textDecoder = new TextDecoder('utf-8', { fatal: false });

export function encodeUtf8(payload: string): Uint8Array {
    return textEncoder.encode(payload);
}

export function decodeUtf8(bytes: Uint8Array): string {
    return textDecoder.decode(bytes);
}

/** Renders one payload as a length-prefixed frame. Throws when the payload
 * exceeds MAX_FRAME_BYTES (callers validate before encoding). */
export function makeFrame(payload: string | Uint8Array): Uint8Array {
    const bytes = typeof payload === 'string' ? encodeUtf8(payload) : payload;
    if (bytes.byteLength > MAX_FRAME_BYTES) {
        throw new RangeError(
            `frame payload of ${bytes.byteLength} bytes exceeds the ${MAX_FRAME_BYTES} byte cap`,
        );
    }
    const frame = new Uint8Array(FRAME_HEADER_BYTES + bytes.byteLength);
    const view = new DataView(frame.buffer);
    view.setUint32(0, bytes.byteLength, true);
    frame.set(bytes, FRAME_HEADER_BYTES);
    return frame;
}

export type FrameExtract =
    | { status: 'need-more-data' }
    | { status: 'message'; message: Uint8Array; rest: Uint8Array }
    | { status: 'protocol-error'; reason: string };

/** Extracts one complete frame from the front of `buffer`. Stateless, shared
 * by both ends; a protocol error means the connection must be closed. */
export function tryExtractFrame(buffer: Uint8Array): FrameExtract {
    if (buffer.byteLength < FRAME_HEADER_BYTES) {
        return { status: 'need-more-data' };
    }
    const view = new DataView(buffer.buffer, buffer.byteOffset, buffer.byteLength);
    const length = view.getUint32(0, true);
    if (length > MAX_FRAME_BYTES) {
        return {
            status: 'protocol-error',
            reason: `declared frame length ${length} exceeds the ${MAX_FRAME_BYTES} byte cap`,
        };
    }
    if (buffer.byteLength < FRAME_HEADER_BYTES + length) {
        return { status: 'need-more-data' };
    }
    const end = FRAME_HEADER_BYTES + length;
    return {
        status: 'message',
        message: buffer.slice(FRAME_HEADER_BYTES, end),
        rest: buffer.slice(end),
    };
}

File: runtime/ipc/src/framing.cpp
#include <mirage/runtime/ipc/framing.hpp>

#include <cstdint>

namespace mirage::runtime::ipc {

std::string make_frame(std::string_view payload) {
    std::string frame;
    frame.reserve(kFrameHeaderBytes + payload.size());
    const std::uint32_t length = static_cast<std::uint32_t>(payload.size());
    frame.push_back(static_cast<char>(length & 0xFFu));
    frame.push_back(static_cast<char>((length >> 8) & 0xFFu));
    frame.push_back(static_cast<char>((length >> 16) & 0xFFu));
    frame.push_back(static_cast<char>((length >> 24) & 0xFFu));
    frame.append(payload.data(), payload.size());
    return frame;
}

FrameExtraction try_extract_frame(std::string& buffer) {
    if (buffer.size() < kFrameHeaderBytes) {
        return {FrameExtract::NeedMoreData, {}, {}};
    }
    const auto byte_at = [&buffer](std::size_t index) {
        return static_cast<std::uint8_t>(buffer[index]);
    };
    const std::uint32_t length = static_cast<std::uint32_t>(byte_at(0)) |
                                 (static_cast<std::uint32_t>(byte_at(1)) << 8) |
                                 (static_cast<std::uint32_t>(byte_at(2)) << 16) |
                                 (static_cast<std::uint32_t>(byte_at(3)) << 24);
    if (length > kMaxFrameBytes) {
        return {FrameExtract::ProtocolError, {},
                "declared frame length " + std::to_string(length) +
                    " exceeds the " + std::to_string(kMaxFrameBytes) +
                    " byte cap"};
    }
    if (buffer.size() < kFrameHeaderBytes + static_cast<std::size_t>(length)) {
        return {FrameExtract::NeedMoreData, {}, {}};
    }
    FrameExtraction extraction;
    extraction.status = FrameExtract::Message;
    extraction.message.assign(buffer, kFrameHeaderBytes,
                              static_cast<std::size_t>(length));
    buffer.erase(0, kFrameHeaderBytes + static_cast<std::size_t>(length));
    return extraction;
}

} // namespace mirage::runtime::ipc
